desc/magnetic_fields.py

```diff
@@ -1094,3 +1130,644 @@ def odefun(rpz, s):
+            rs = rpz2xyz_vec(_rs, phi=phi)
+            K = rpz2xyz_vec(_K, phi=phi)
+            fj = biot_savart_general(
+                coords,
+                rs,
+                K,
+                _dV,
+            )
+            f += fj
+            return f
+
+        B = fori_loop(0, self.surface_grid.NFP, nfp_loop, jnp.zeros_like(coords))
+        if basis == "rpz":
+            B = xyz2rpz_vec(B, x=coords[:, 0], y=coords[:, 1])
+        return B
+
+    @classmethod
+    def from_surface(
+        cls,
+        surface,
+        potential,
+        potential_dtheta,
+        potential_dzeta,
+        surface_grid,
+        params=None,
+        rho=1,
+        name="",
+        check_orientation=True,
+    ):
+        """Create CurrentPotentialField using geometry provided by given surface.
+
+        Parameters
+        ----------
+        surface: FourierRZToroidalSurface, optional, default None
+            Existing FourierRZToroidalSurface object to create a
+            CurrentPotentialField with.
+        potential : callable
+            function to compute the current potential. Should have a signature of
+            the form potential(theta,zeta,**params) -> ndarray.
+            theta,zeta are poloidal and toroidal angles on the surface
+        potential_dtheta: callable
+            function to compute the theta derivative of the current potential
+        potential_dzeta: callable
+            function to compute the theta derivative of the current potential
+        surface_grid : Grid,
+            grid upon which to evaluate the surface current density K
+        params : dict, optional
+            default parameters to pass to potential function (and its derivatives)
+
+        name : str
+            name for this field
+        check_orientation : bool
+            ensure that this surface has a right handed orientation. Do not set to False
+            unless you are sure the parameterization you have given is right handed
+            (ie, e_theta x e_zeta points outward from the surface).
+
+        """
+        if not isinstance(surface, FourierRZToroidalSurface):
+            raise TypeError(
+                "Expected type FourierRZToroidalSurface for argument surface, "
+                f"instead got type {type(surface)}"
+            )
+        R_lmn = surface.R_lmn
+        Z_lmn = surface.Z_lmn
+        modes_R = surface._R_basis.modes[:, 1:]
+        modes_Z = surface._Z_basis.modes[:, 1:]
+        NFP = surface.NFP
+        sym = surface.sym
+        rho = surface.rho
+
+        return cls(
+            potential,
+            potential_dtheta,
+            potential_dzeta,
+            surface_grid,
+            params,
+            R_lmn,
+            Z_lmn,
+            modes_R,
+            modes_Z,
+            NFP,
+            sym,
+            rho,
+            name,
+            check_orientation,
+        )
+
+
+class FourierCurrentPotentialField(_MagneticField, FourierRZToroidalSurface):
```

Comment:
The reason why this is not subclassed from `CurrentPotentialField` is that I wanted to be able to have the compute function for `Phi` from this use the transforms and `Phi_basis` cleanly, in that it will get a transform with `get_transforms` and see the `Phi_basis` and use that in the transform. 

That requires separate compute functions for the `Phi` from this `FourierCurrentPotentialField` class and the `CurrentPotentialField` class. However, if the fourier was a subclass of the  `CurrentPotentialField`, our API for compute gets confused as it thinks I've defined the same function twice for the same parametrization. 

I was unable to think of how to avoid this without just separating the two classes. This works OK as the `K` compute variable only uses the calculated derivatives and geometry, so that can have one entry for both classes.

The only unclean part is that I now have duplicated code in the `compute_magnetic_field` function for these two classes, as they are the exact same.

Maybe a possible solution is to make these 2 separate subclasses of an overarching `SurfaceCurrentPotentialField`, but that would only be to avoid duplicating the compute_magnetic_field code